File: Python/Container With Most Water.py
# Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

# Note: You may not slant the container
class Solution:
    # @return an integer
    def maxArea(self, height):
		lenH = len(height)
		contain = 0
		if lenH < 2:
			return 0
		end = lenH-1
		maxV=0
		i=0
		while i<end:
			contain = min(height[i],height[end])*(end - i)
			maxV = max(maxV,contain)
			if height[i] <= height[end]:
				i = i + 1
			else:
				end = end -1
		return maxV

File: Java/Container With Most Water.java
class Container With Most Water{
	public int maxArea(int[] height) {
        int minHeight = 0;
        int max = 0;
        int left = 0;
        int right = height.length-1;
        while(left < right){
        	int curr = Math.min(height[left],height[right])*(right-left);
        	max = Math.max(max,curr);
        	int currL = left;
        	int currR = right;
        	if(height[left]<height[right]){
        		left++;
        		while(left<right && height[left]<=height[currL])
        			left++;
        	}
        	else{
        		right--;
        		while(left < right && height[right]<=height[currR])
        			right--;
        	}
        }
        return max;
    }
}